tools/audio/soundfont_compiler.c

```diff
@@ -0,0 +1,1811 @@
+
+    document = xmlReadFile(filename_in, NULL, XML_PARSE_NONET);
+    if (document == NULL)
+        return EXIT_FAILURE;
+
+    xmlNodePtr root = xmlDocGetRootElement(document);
+    if (!strequ(XMLSTR_TO_STR(root->name), "Soundfont"))
+        error("Root node must be <Soundfont>");
+    read_soundfont_info(&sf, root);
+
+    sf.envelopes = sf.envelope_last = NULL;
+
+    // read all envelopes first irrespective of their positioning in the xml
+    LL_FOREACH(xmlNodePtr, node, root->children) {
+        const char *name = XMLSTR_TO_STR(node->name);
+
+        if (strequ(name, "Envelopes"))
+            read_envelopes_info(&sf, node);
+    }
+
+    // read all samples
+    sf.samples = NULL;
+    LL_FOREACH(xmlNodePtr, node, root->children) {
+        const char *name = XMLSTR_TO_STR(node->name);
+
+        if (strequ(name, "Samples"))
+            read_samples_info(&sf, node);
+    }
+
+    // read all instruments
+    memset(sf.program_number_bitset, 0, sizeof(sf.program_number_bitset));
+    sf.instruments = NULL;
+    sf.drums = NULL;
+    sf.sfx = NULL;
+    LL_FOREACH(xmlNodePtr, node, root->children) {
+        const char *name = XMLSTR_TO_STR(node->name);
+
+        if (strequ(name, "Instruments"))
+            read_instrs_info(&sf, node);
+        if (strequ(name, "Drums"))
+            read_drums_info(&sf, node);
+        if (strequ(name, "Effects"))
+            read_sfx_info(&sf, node);
+    }
+
+    // read match padding if it exists
+    sf.match_padding = NULL;
+    LL_FOREACH(xmlNodePtr, node, root->children) {
+        const char *name = XMLSTR_TO_STR(node->name);
+
+        if (strequ(name, "MatchPadding"))
+            read_match_padding(&sf, node);
+    }
+
+    // emit C source
+
+    FILE *out_c = fopen(filename_out_c, "w");
+    fprintf(out_c, "#include \"soundfont_file.h\"\n\n");
+
+    size_t size = 0;
+    size += emit_c_header(out_c, &sf);
+    size += emit_c_samples(out_c, &sf);
+    size += emit_c_envelopes(out_c, &sf);
+    size += emit_c_instruments(out_c, &sf);
+    size += emit_c_drums(out_c, &sf);
+    size += emit_c_effects(out_c, &sf);
+    emit_c_match_padding(out_c, &sf, size);
+
+    fclose(out_c);
+
+    // emit C header
+
+    FILE *out_h = fopen(filename_out_h, "w");
+    fprintf(out_h,
+            // clang-format off
+           "#ifndef SOUNDFONT_%d_H_"            "\n"
+           "#define SOUNDFONT_%d_H_"            "\n"
+                                                "\n",
+            // clang-format on
+            sf.info.index, sf.info.index);
+
+    fprintf(out_h,
+            // clang-format off
+           "#ifdef _LANGUAGE_ASEQ"              "\n"
+           ".pushsection .fonts, \"\", @note"   "\n"
+           "    .byte %d /*sf id*/"             "\n"
+           ".popsection"                        "\n"
+           "#endif"                             "\n"
+                                                "\n",
+            // clang-format on
+            sf.info.index);
```

Comment:
what is `_LANGUAGE_ASEQ`?

Reply:
It's for sequence assembly files that are yet to be introduced. Sequence assembly files are preprocessed and `_LANGUAGE_ASEQ` is pre-defined for them (analogous to `_LANGUAGE_C` and `_LANGUAGE_ASSEMBLY`)